### ISSUE TITLE
Passing arguments to Jou programs

### ISSUE THREAD

**Issue by Akuli** [Akuli/jou#257]
Currently not possible, unless you first compile to a file with `jou -o filename.exe program.jou` and then pass arguments like so `./filename.exe arg1 arg2 arg3`

**Comment by littlewhitecloud** [comment]
we don't need passing arguments to Jou programs when compile. There is no programming language gives args to programs when compile

**Comment by Akuli** [comment]
Indeed. If you use `-o`, it causes Jou to only compile the program and not run it, so it shouldn't accept additional arguments if `-o` is used. But without `-o` it compiles and then runs the program, and that's what this issue is about.

**Comment by Moosems** [comment]
> we don't need passing arguments to Jou programs when compile. There is no programming language gives args to programs when compile

And why can't Jou be the first 😄? I think it could actually have some amazing applications.

**Comment by Akuli** [comment]
C compilers have the `-D` option, which basically specifies a compile-time argument. If you pass `-Dfoo=123` to the compiler, then every occurrence of the variable name `foo` in the `.c` files will be replaced by 123, as if the C code contained `#define foo 123`.

I think Jou should have something similar, but it is somewhat low priority. You could instead auto-generate a file that gets imported into Jou. In fact, I already do that:

https://github.com/Akuli/jou/blob/b709670adaad11dec65b742716f6dd0c78b57015/Makefile.posix#L33-L36

**Comment by Moosems** [comment]
Is there another way besides just overriding every occurrence of a variable?

**Comment by Akuli** [comment]
In C there really isn't, and with properly named variables, it isn't as bad as you would expect it to be.

For Jou, the best option would probably be reading environment variables at compile time. Environment variables are a "standard" way to pass info to programs that most developers are already familiar with (and if not, will be very useful to learn). This way compiling the program would look something like `MY_VAR=123 jou program.jou`.

**Comment by Moosems** [comment]
Hmmm. The reason I ask is because I could imagine myself wanting to use compile-time optimizations for code that is dependent one one thing or another such as platform or whatnot that can be determined dynamically in the code when speed isn't a priority but can be optimized for distribution.

**Comment by Akuli** [comment]
You can already do `if WINDOWS:` or `if MACOS:` at compile time. That will probably need to be generalized at some point in the future.

**Comment by Moosems** [comment]
I know, I was just using that as an example.

**Comment by Akuli** [comment]
I think this should be a compiler flag, like this:

```
jou filename.jou --args "arg1 arg2 arg3"
```

This would be basically same as:

```
jou -o program filename.jou
./program arg1 arg2 arg3
```
